histogram/src/standard.rs

```diff
@@ -130,6 +130,41 @@ impl Histogram {
         }
     }
 
+    /// Returns a new histogram with a reduced grouping power.
+    ///
+    /// The specified factor determines how much the grouping power is reduced
+    /// by and should be smaller than the histogram's existing grouping power.
+    /// Every step of grouping power halves the total number of buckets (and
+    /// hence total size of thie histogram), while doubling the error.
+    ///
+    /// This works by iterating over every bucket in the existing histogram
+    /// and inserting the contained values into the new histogram. Since we
+    /// do not know the exact values of the data points (only that they lie
+    /// within the bucket's range), we pick a pessimistic high value.
+    pub fn downsample(&self, factor: u8) -> Result<Histogram, BuildError> {
```

Comment:
I believe we can do this on the raw slices and avoid the index->value value->index conversions. Imagine the simple case where grouping power was 1 and now we want to lower it to 0, and let's say the max value power was 3:

```
┌───────┬───────┬───────┬───────┬───────────────┬───────────────────────┐
│       │       │       │       │               │                       │
│   0   │   1   │   2   │   3   │     4 - 5     │         6 - 8         │
│       │       │       │       │               │                       │
└───────┴───────┴───────┴───────┴───────────────┴───────────────────────┘
                                                                         
┌───────┬───────┬───────────────┬───────────────────────────────────────┐
│       │       │               │                                       │
│   0   │   1   │     2 - 3     │                 4 - 8                 │
│       │       │               │                                       │
└───────┴───────┴───────────────┴───────────────────────────────────────┘
```

Reply:
You can determine the position to start merging by taking `2^(new grouping power) + 1` and merge bucket counts pairwise

Reply:
Per discussion, we can follow-up on this. For now the double conversion shouldn't be a bottleneck in any critical path. We can revisit if we experience performance issues.